rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs

```diff
@@ -0,0 +1,249 @@
+}
+
+fn construct_flat_fact(ctx: &FlatFactContextAll) -> AstFlatFact {
+    match ctx {
+        FlatFactContextAll::PrinFactContext(fctx) => construct_prin_fact(fctx),
+        FlatFactContextAll::PredFactContext(pctx) => construct_pred_fact(pctx),
+        _ => {
+            panic!("construct_flat_fact tried to build error");
+        }
+    }
+}
+
+fn construct_prin_fact(ctx: &PrinFactContext) -> AstFlatFact {
+    let prin = construct_principal(&ctx.principal().unwrap());
+    let verbphrase = construct_verbphrase(&ctx.verbphrase().unwrap());
+    AstFlatFact::AstPrinFact {
+        p: prin,
+        v: verbphrase,
+    }
+}
+
+fn construct_pred_fact(ctx: &PredFactContext) -> AstFlatFact {
+    let pred = construct_predicate(&ctx.predicate().unwrap());
+    AstFlatFact::AstPredFact { p: pred }
+}
+
+fn construct_fact(ctx: &FactContextAll) -> AstFact {
+    match ctx {
+        FactContextAll::FlatFactFactContext(fctx) => construct_flat_fact_fact(fctx),
+        FactContextAll::CanSayFactContext(sctx) => construct_can_say_fact(sctx),
+        _ => {
+            panic!("construct_fact tried to build error");
+        }
+    }
+}
+
+fn construct_flat_fact_fact(ctx: &FlatFactFactContext) -> AstFact {
+    let flat = construct_flat_fact(&ctx.flatFact().unwrap());
+    AstFact::AstFlatFactFact { f: flat }
+}
+
+fn construct_can_say_fact(ctx: &CanSayFactContext) -> AstFact {
+    let prin = construct_principal(&ctx.principal().unwrap());
+    let fact = construct_fact(&ctx.fact().unwrap());
+    AstFact::AstCanSayFact {
+        p: prin,
+        f: Box::new(fact),
+    }
+}
+
+fn construct_assertion(ctx: &AssertionContextAll) -> AstAssertion {
+    match ctx {
+        AssertionContextAll::FactAssertionContext(fctx) => construct_fact_assertion(fctx),
+        AssertionContextAll::HornClauseAssertionContext(hctx) => construct_hornclause(hctx),
+        _ => {
+            panic!("construct_assertion tried to build error");
+        }
+    }
+}
+
+fn construct_fact_assertion(ctx: &FactAssertionContext) -> AstAssertion {
+    let fact = construct_fact(&ctx.fact().unwrap());
+    AstAssertion::AstFactAssertion { f: fact }
+}
+
+fn construct_hornclause(ctx: &HornClauseAssertionContext) -> AstAssertion {
+    let lhs = construct_fact(&ctx.fact().unwrap());
+    let mut rhs = Vec::new();
+    for flat_fact_ctx in ctx.flatFact_all() {
+        rhs.push(construct_flat_fact(&flat_fact_ctx));
+    }
+    AstAssertion::AstCondAssertion { lhs, rhs }
+}
+
+fn construct_says_assertion(ctx: &SaysAssertionContextAll) -> AstSaysAssertion {
+    match ctx {
+        SaysAssertionContextAll::SaysSingleContext(ctx_prime) => {
+            let prin = construct_principal(&ctx_prime.principal().unwrap());
+            let assertions = vec![construct_assertion(&ctx_prime.assertion().unwrap())];
+            let export_file: Option<String> = match ctx_prime.PATH() {
+                Some(_) => Some(ctx_prime.PATH().unwrap().get_text()),
```

Comment:
Is there a reason for not using something like: `Some(path) => Some(path.get_text())` or replacing the entire match expression with https://doc.rust-lang.org/std/option/enum.Option.html#method.map?

Reply:
Using map. Suggestion applied broadly.